### PR TITLE
Allow Invoice Refunding

### PR DESCRIPTION
**Pull request by sampoder**
Fixes #4484

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -259,6 +259,9 @@
     post "unarchive"
     get "hosted"
     get "pdf"
+    member do
+      post "refund", to: "invoices#refund"
```
Comment (cjdenio) [inline config/routes.rb]:
```suggestion
      post "refund"
```
handy shorthand

File: app/models/invoice.rb [hunk 2]
```diff
@@ -144,13 +145,18 @@ class NoAssociatedStripeCharge < StandardError; end
...
+    end
   end
 
   enum status: {
     draft: "draft", # only 3 invoices [203, 204, 128] leftover from when drafts existed
     open: "open",
     paid: "paid",
-    void: "void"
+    void: "void",
+    refunded: "refunded",
```
Comment (cjdenio) [inline app/models/invoice.rb]: The `status` field tracks the [Stripe invoice status](https://stripe.com/docs/api/invoices/object#invoice_object-status), and shouldn't include values besides the ones in the documentation

File: app/services/invoice_service/refund.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        # 1. Mark refunded
+        invoice.mark_refunded!
+
+        # 2. Un-front all pending transaction associated with this donation
+        invoice.canonical_pending_transactions.update_all(fronted: false)
+
+        # 3. Process remotely
+        ::StripeService::Refund.create(charge: stripe_charge_id, amount: @amount)
+
+        # 4. Create top-up on Stripe. Located in `StripeController#handle_charge_refunded`
```
Comment (cjdenio) [inline app/services/invoice_service/refund.rb]: Looks like StripeController still needs to be updated to handle invoices

**Review by cjdenio — DISMISSED** [review]

File: config/routes.rb (hunk 1 above)
Comment (sampoder) [reply config/routes.rb]: nice!

File: app/models/invoice.rb (hunk 2 above)
Comment (sampoder) [inline app/models/invoice.rb]:
```suggestion
```

File: app/services/invoice_service/refund.rb (hunk 3 above)
Comment (sampoder) [reply app/services/invoice_service/refund.rb]: mb, that should be in there now!

**Comment by sampoder** [comment]
@cjdenio: thoughts on using the `void` state? feels appropriate to me but I'm not quite sure

File: app/controllers/stripe_controller.rb [hunk 4]
```diff
@@ -169,16 +169,27 @@ def handle_charge_refunded(event)
...
+      donation.mark_refunded! unless donation.refunded?
+
+      StripeService::Topup.create(
+        amount: charge[:amount_refunded],
+        currency: "usd",
+        description: "Refund for donation #{donation.id}",
+        statement_descriptor: "HCB-#{donation.local_hcb_code.short_code}"
+      )
+    elsif invoice
+      invoice.mark_void! unless donation.void?
```
Comment (sampoder) [inline app/controllers/stripe_controller.rb]:
```suggestion
      invoice.mark_void! unless invoice.void?
```

sleepy me

**Comment by cjdenio** [comment]
I think it can stay `paid`, no? We just use the `aasm_state` field to remember that it was refunded after the fact

**Comment by sampoder** [comment]
Oh ok cool, I'd assumed `aasm_state` has had to match it's Stripe status

File: app/controllers/stripe_controller.rb [hunk 5]
```diff
@@ -169,16 +169,27 @@ def handle_charge_refunded(event)
...
+      donation.mark_refunded! unless donation.refunded?
+
+      StripeService::Topup.create(
+        amount: charge[:amount_refunded],
+        currency: "usd",
+        description: "Refund for donation #{donation.id}",
+        statement_descriptor: "HCB-#{donation.local_hcb_code.short_code}"
+      )
+    elsif invoice
+      invoice.mark_refunded! unless invoice.refunded?
```
Comment (sampoder) [inline app/controllers/stripe_controller.rb]:
```suggestion
      invoice.mark_refunded! unless invoice.refunded_v2?
```

Not quite sure how AASM state works, would this make sense?

File: app/services/invoice_service/refund.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  class Refund
+    def initialize(invoice_id:, amount:)
+      @invoice_id = invoice_id
+      @amount = amount
+    end
+
+    def run
+      ActiveRecord::Base.transaction do
+
+        # 1. Un-front all pending transaction associated with this donation
```
Comment (sampoder) [inline app/services/invoice_service/refund.rb]:
```suggestion
        # 1. Un-front all pending transaction associated with this invoice
```

**Comment by sampoder** [comment]
https://hackclub.slack.com/archives/C047Y01MHJQ/p1715204152953299

**Pull request merged**